### PR TITLE
2371_2

### PR DESCRIPTION
**Pull request by brutu33**
## Номер
2371

## Номер задания
2

## Ссылка на видео с демо
https://drive.google.com/open?id=1bCtNwBxyZoW4tWTM7hijnE-KzTkO7hxE

## Комментарии
Finish task 2: Done searching authors and all info about them (battles count, curse words and etc). Output info show in table

### CODE REVIEW

File: 2371/2/application.rb [hunk 1]
```diff
@@ -0,0 +1,87 @@
+require 'optparse'
+require File.expand_path(File.dirname(__FILE__) + '/rap_battles_manager')
```
Comment (alexshgov) [inline 2371/2/application.rb]: You could just use `require_relative`

File: 2371/2/application.rb [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+require File.expand_path(File.dirname(__FILE__) + '/constants')
+
+# The Application responsible for start searching info
+class Application
+  attr_reader :options
+  def initialize
+    @thread = Thread.new do
```
Comment (alexshgov) [inline 2371/2/application.rb]: Ideally, in an initializer, you should set only the initial state.
All further logic will come after you send the appropriate message to your object.
I suggest you remove this loading functionality.

File: 2371/2/application.rb [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+      0.step(1000, 5) do |step|
+        printf("\rSearching:  %-20s", '*' * (step / 5))
+        sleep(0.5)
+      end
+    end
+    @manager = RapBattlesManager.new
+    @options = {}
+  end
+
+  def show_authors_bad_w_rating(number)
```
Comment (alexshgov) [inline 2371/2/application.rb]: It took some time to figure out that `w` is `word`. Please, be explicit in naming.

File: 2371/2/application.rb [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+unless File.directory?('./texts/')
+  puts NO_DIRECTORY
+  exit
+end
+
+ARGV << '-h' if ARGV.empty?
+
+@app = Application.new
+
+option_parser = OptionParser.new do |opts|
```
Comment (alexshgov) [inline 2371/2/application.rb]: **The rule: one class -- one file.**
Move this stuff out of this file.

File: 2371/2/author_entity.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
+require File.expand_path(File.dirname(__FILE__) + '/constants')
+require File.expand_path(File.dirname(__FILE__) + '/battle_entity')
+
+# AuthorEntity responsible for author info
+class AuthorEntity
```
Comment (alexshgov) [inline 2371/2/author_entity.rb]: You could just name it as `Author`. No need to add Entity word.
OOP is about business entities. When you create `Author` class it's clear that this class will represent an author.

File: 2371/2/author_entity.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  attr_reader :author_name, :battles_info
+  def initialize(author_name, battle_name)
+    @author_name = author_name
+    @battles = []
+    @battles_info = { curse_words: 0, words_percent: 0 }
+    add_battle(battle_name)
```
Comment (alexshgov) [inline 2371/2/author_entity.rb]: Do not place any logic in the initializer. Just set the initial state here.

File: 2371/2/author_entity.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
+require File.expand_path(File.dirname(__FILE__) + '/constants')
+require File.expand_path(File.dirname(__FILE__) + '/battle_entity')
+
+# AuthorEntity responsible for author info
+class AuthorEntity
+  attr_reader :author_name, :battles_info
+  def initialize(author_name, battle_name)
```
Comment (alexshgov) [inline 2371/2/author_entity.rb]: I don't understand, why the `Author` class should be coupled to battle_name?

File: 2371/2/author_entity.rb [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    add_battle(battle_name)
+  end
+
+  def add_battle(battle_file_name)
+    return unless File.file?("./texts/#{battle_file_name}")
+    @battles << BattleEntity.new(battle_file_name)
+    prepare_data
+  end
+
+  def bad_words_print
```
Comment (alexshgov) [inline 2371/2/author_entity.rb]: Treat your class like YOU are the user of the class. Thinks about **messages** which you pass to the object.
`Hey author, print me bad words!`
`Ok! author.print_bad_words`

But anyway, I don't like the idea that Author knows something about **Formatting**.
Ideally, your Author shouldn't know anything about formatting.

```ruby
class Author
  def bad_words; end # returns an array of bad words
  def bad_words_per(battle); end # returns Hash like { battle_1 => bad_words_count, battle_2 => ... }
end
```

File: 2371/2/battle_entity.rb [hunk 9]
```diff
@@ -0,0 +1,22 @@
+# The BattleEntity responsible for battle info
+class BattleEntity
+  def initialize(file_name)
+    @curse_words = []
+    @words_percent = 0
+    manage_battle_text(File.read("./texts/#{file_name}"))
```
Comment (alexshgov) [inline 2371/2/battle_entity.rb]: No logic in initializers!
Think about the messages which you will pass to the `Battle` object.

My suggestion

```ruby
# battle.rb
class Battle
  attr_reader :battle_file_path, ... #etc
  
  def intialize(battle_file_path)
     # ...
  end

  def words_count; end
  def rounds_count; end
  def bad_words_count; end
  def battler_name; end
  # ...
end
```

**Comment by alexshgov** [comment]
Let's start with the comments above.

Moreover, I see that you have done the task in one commit.
It is unacceptable! In the future, I want to see **atomic and reasonable commits**. Thanks.

File: 2371/2/author.rb [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+  def add_battle(battle)
+    @battles << battle
+  end
+
+  def bad_words
+    @battles.map(&:curse_words).flatten
+  end
+
+  def words_percent
+    @battles.map(&:words_percent).inject(0, &:+)
```
Comment (allomov) [inline 2371/2/author.rb]: `bad_words_percent `

File: 2371/2/content_manager.rb [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def add_battle_to_author(new_author, battle)
+    unless @authors.detect do |author|
+      next unless author.name?(new_author.name)
+      author.add_battle(battle)
+      break author
+    end
+      new_author.add_battle(battle)
+      @authors << new_author
```
Comment (allomov) [inline 2371/2/content_manager.rb]: "закинь" поиск/создание артиста в другой класс (например rappers_registry).

File: 2371/2/content_manager.rb [hunk 12]
```diff
@@ -0,0 +1,42 @@
+require_relative 'author'
+# The ContentManager responsible for searching authors info in files
+class ContentManager
```
Comment (allomov) [inline 2371/2/content_manager.rb]: получается, что ContentManager и читает и выводит данные на экран, лучше вынести вывод в отдельный класс

File: 2371/2/author.rb [hunk 13]
```diff
@@ -20,19 +20,19 @@ def bad_words
...
-  def words_percent
-    @battles.map(&:words_percent).inject(0, &:+)
+  def bad_words_per_battles
+    bad_words.size.to_f / @battles.size
   end
 
-  def words_per_battle
-    bad_words.size.to_f / @battles.size
+  def words_per_battles_rounds
+    @battles.map(&:words_per_round).inject(0, &:+)
```
Comment (allomov) [inline 2371/2/author.rb]: получается хорошо. двигайся в том же направлении.

**Comment by brutu33** [comment]
Made changes in project. Moved print methods from rap_battle_controller.rb and author.rb to classes for print. Move progress to class. Add method for search author by name.

File: 2371/2/author.rb [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+  def bad_words
+    @battles.map(&:curse_words).flatten
+  end
+
+  def bad_words_per_battles
+    bad_words.size.to_f / @battles.size
+  end
+
+  def words_per_battles_rounds
+    @battles.map(&:words_per_round).inject(0, &:+)
```
Comment (allomov) [inline 2371/2/author.rb]: тут не нужно их делить? это немного странная статистика, назови методы лучше

File: 2371/2/authors_content_manager.rb [hunk 15]
```diff
@@ -0,0 +1,42 @@
+require_relative 'author'
+require_relative 'constants'
+require 'set'
+# The AuthorsContentManager responsible for manage authors
+class AuthorsContentManager
+  def initialize
+    @authors_set = Set.new
+  end
+
+  def add?(content)
```
Comment (allomov) [inline 2371/2/authors_content_manager.rb]: неправильно назван метод

File: 2371/2/authors_content_manager.rb [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    author = find_author_by(name) || Author.new(name)
+    @authors_set.add? author
+    author
+  end
+
+  def find_author_by(name)
+    unless @authors_set.detect do |author|
+      next unless author.name?(name)
+      return author
+    end
```
Comment (allomov) [inline 2371/2/authors_content_manager.rb]: это работает?

File: 2371/2/authors_content_manager.rb [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  private
+
+  def find_or_create_author_by_name(name)
+    author = find_author_by(name) || Author.new(name)
+    @authors_set.add? author
+    author
+  end
+
+  def find_author_by(name)
+    unless @authors_set.detect do |author|
```
Comment (allomov) [inline 2371/2/authors_content_manager.rb]: можно лучше написать

File: 2371/2/authors_content_manager.rb [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+
+  def add?(content)
+    raise NO_CONTENT_WARN unless content.any?
+    content.each do |info|
+      find_or_create_author_by_name(info[:name]).add_battle(Battle.new(info[:battle_text]))
+    end
+  end
+
+  def authors
+    raise NO_AUTHOR_WARN unless @authors_set.to_a.any?
```
Comment (allomov) [inline 2371/2/authors_content_manager.rb]: просто пустой массив вернуть?

File: 2371/2/authors_content_manager.rb [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+    @authors_set
+  end
+
+  def author_with_name(name)
+    raise NO_AUTHOR_WARN unless @authors_set.to_a.any?
```
Comment (allomov) [inline 2371/2/authors_content_manager.rb]: WARN != Exception

File: 2371/2/battle.rb [hunk 20]
```diff
@@ -0,0 +1,18 @@
+require_relative 'constants'
+
+# The Battle responsible for battle info
+class Battle
+  attr_reader :curse_words
+  def initialize(battle_text)
+    @battle_text = battle_text
+    all_words = @battle_text.scan(/[а-яА-Я]+/i)
+    @curse_words = (all_words - (all_words - CURSE_WORDS_DICT)) \
```
Comment (allomov) [inline 2371/2/battle.rb]: может в методе это делать, а не при инициализации?

**Comment by brutu33** [comment]
battle.rb - move finding badwords to method. author_content_manager.rb - clear code. Rename methods in classes. Change raise warning about empty content and empty authors array to puts.

**Comment by alexshgov** [comment]
I don't like the idea of having classes like `*Controller` and `*Manager` (forget about Rails for a second) because they represent, actually, nothing.
Instead of `*Manager` we could use something like `*Registry`.

`*Controller` is a little bit sophisticated one, I see "_The RapBattlesController responsible for start searching info_" description. So, if it's a Search, why not use the following interface: `RapBattlesSearch.new(args = {}).bad_words_rating` and so on?

My final grade for the work is **5/7**

@allomov if you have any further comments, please, feel free to add them.

**Comment by allomov** [comment]
I agree that `Controller` concept is complicated. I expected there were some readers classes there.

Сoncerning `Manager` vs `Registry` - it's better to have `Manager` that can be called `Registry`, than nothing.

**Comment by allomov** [comment]
название PR )

**Comment by allomov** [comment]
название commitов

File: 2371/2/authors_info_searcher.rb [hunk 21]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  def authors_bad_words_rating
+    @authors.sort_by { |author| -author.bad_words.size }
+  end
+
+  private
+
+  def author_with_name(name)
+    @authors.detect { |author| break author if SimilarItemsSearcher.compare_items(name, author.name) }
```
Comment (allomov) [inline 2371/2/authors_info_searcher.rb]: можно проще сделать

File: 2371/2/battle.rb [hunk 22]
```diff
@@ -0,0 +1,19 @@
+require_relative 'constants'
+# The Battle responsible for battle info
+class Battle
+  def initialize(battle_text)
+    @battle_text = battle_text
+  end
+
+  def bad_words
+    all_words = @battle_text.scan(/[а-яА-Я]+/i)
```
Comment (allomov) [inline 2371/2/battle.rb]: отдельный метод

File: 2371/2/battle.rb [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+    star_words = @battle_text.scan(/\W[а-яА-Я]*\*+[а-яА-Я]*\W/i).flatten
+    (all_words - (all_words - CURSE_WORDS_DICT)) + star_words
+  end
+
+  def words_per_round
```
Comment (allomov) [inline 2371/2/battle.rb]: string or float?

**Comment by allomov** [comment]
changes are approved, everything is done as expect, mentor is satisfied :)

**Pull request merged**